drivers/sensor/stm32_vbat/stm32_vbat.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2022 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <drivers/sensor.h>
+#include <drivers/adc.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(stm32_vbat, CONFIG_SENSOR_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_vbat)
+#define DT_DRV_COMPAT st_stm32_vbat
+#else
+#error "No compatible devicetree node found"
+#endif
+
+struct stm32_vbat_data {
+	const struct device *adc;
+	const struct adc_channel_cfg adc_cfg;
+	struct adc_sequence adc_seq;
+	struct k_mutex mutex;
+	int16_t sample_buffer;
+	int16_t raw; /* raw adc Sensor value */
+};
+
+struct stm32_vbat_config {
+	int vbat_mv;
+	int ratio;
+};
+
+static int stm32_vbat_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct stm32_vbat_data *data = dev->data;
+	struct adc_sequence *sp = &data->adc_seq;
+	int rc;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_VOLTAGE) {
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	rc = adc_channel_setup(data->adc, &data->adc_cfg);
+
+	if (rc) {
+		LOG_DBG("Setup AIN%u got %d", data->adc_cfg.channel_id, rc);
+		return rc;
```

Comment:
@FRASTM, but the mutex is not unlocked.

```suggestion
		k_mutex_unlock(&data->mutex);
		return rc;
```

Reply:
using ` goto: ` like the in the _drivers/sensor/stm32_temp/stm32_temp.c_